### PR TITLE
Cleans up a lot of code/joysticks/gamepads

### PR DESCRIPTION
**Pull request by RobertBColton**
Fixes formatting on the joystick functions and makes the gamepad functions behave more like GMS. Also uses an improved method to ENIGMA.exe for finding the installed Java.

Uses [^\n]\z to find missing newlines and ( |\t)+\r to find excess whitespace.

### CODE REVIEW

File: ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSjoystick.cpp [hunk 1]
```diff
@@ -32,108 +32,103 @@ namespace enigma_user
...
+  if (joySetCapture(enigma::hWnd, JOYSTICKID1 + id, 0, FALSE))
+  {
+    MessageBeep(MB_ICONEXCLAMATION);
+    MessageBox(enigma::hWnd, "Couldn't capture the joystick.", NULL,
+        MB_OK | MB_ICONEXCLAMATION);
+    return false;
+  }
+
+  bDevAttached = joyGetPos(JOYSTICKID1 + id, &joyinfo) != JOYERR_UNPLUGGED;
+  if (!bDevAttached)
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSjoystick.cpp]: return !bDevAttached;

File: ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSjoystick.cpp [hunk 2]
```diff
@@ -32,108 +32,103 @@ namespace enigma_user
...
-			return 0;
-	}
+  JOYINFOEX joyinfo;
+  JOYCAPS joycaps;
+  joyGetDevCaps(JOYSTICKID1 + id, &joycaps, sizeof(joycaps));
+  switch (axisnum) {
+    case 1:
+      joyinfo.dwFlags = JOY_RETURNX | JOY_RETURNCENTERED;
+      joyGetPosEx(JOYSTICKID1 + id, &joyinfo);
+      return (joyinfo.dwXpos - (joycaps.wXmax + joycaps.wXmin) / 2.f) / ((joycaps.wXmax - joycaps.wXmin)/ 2.f);
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSjoystick.cpp]: Please extract a `inline float joystick_grade(float position, float min, float max)`.
The body should just be `(pos - (max + min) / 2.f) / ((max - min)/ 2.f);`.
Am I correct in saying that min and max can be declared `constexpr`? Try it.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp [hunk 3]
```diff
@@ -24,231 +23,210 @@
...
-    // Get the state
-    DWORD Result = XInputGetState(num, &controllerState);
-    if (Result == ERROR_SUCCESS) {
-        return true;
-    } else {
-        return false;
-    }
+  XINPUT_STATE controllerState = {};
+  // Get the state
+  DWORD dwResult = XInputGetState(num, &controllerState);
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp]: `return XInputGetState(num, &controllerState) == ERROR_SUCCESS;`

**Comment by RobertBColton** [comment]
I decided to go use a disassembler to figure out where the gamepad_get_description names were coming from and I found that they must be hardcoding them. I can tell because only those asshats would use four exclamation marks to tell you XInput was not found.

```asm
.rdata:006846E4 aDirectinput8cr db 'DirectInput8Create',0 ; DATA XREF: .text:0043E274o
.rdata:006846F7                 align 4
.rdata:006846F8 off_6846F8      dd offset off_690064    ; DATA XREF: .text:0043E252o
.rdata:006846FC                 dd offset unk_70006E
.rdata:00684700                 dd offset unk_740075
.rdata:00684704 a8_dll:
.rdata:00684704                 unicode 0, <8.dll>,0
.rdata:00684710 ; char aXinputenable[]
.rdata:00684710 aXinputenable   db 'XInputEnable',0     ; DATA XREF: sub_43E4A0+60o
.rdata:0068471D                 align 10h
.rdata:00684720 ; char aXinputsetstate[]
.rdata:00684720 aXinputsetstate db 'XInputSetState',0   ; DATA XREF: sub_43E4A0+4Do
.rdata:0068472F                 align 10h
.rdata:00684730 ; char aXinputgetstate[]
.rdata:00684730 aXinputgetstate db 'XInputGetState',0   ; DATA XREF: sub_43E4A0+3Fo
.rdata:0068473F                 align 10h
.rdata:00684740 ; char ProcName[]
.rdata:00684740 ProcName        db 'XInputGetCapabilities',0 ; DATA XREF: sub_43E4A0+2Do
.rdata:00684756                 align 4
.rdata:00684758 ; const WCHAR off_684758
.rdata:00684758 off_684758      dd offset aY_set        ; DATA XREF: sub_43E4A0+17o
.rdata:00684758                                         ; "y_set"
.rdata:0068475C                 dd offset unk_70006E
.rdata:00684760                 dd offset unk_740075
.rdata:00684764                 dd offset loc_5F0031
.rdata:00684768 a4_dll:
.rdata:00684768                 unicode 0, <4.dll>,0
.rdata:00684774 ; const WCHAR LibFileName
.rdata:00684774 LibFileName     dd offset aY_set        ; DATA XREF: sub_43E4A0+7o
.rdata:00684774                                         ; "y_set"
.rdata:00684778                 dd offset unk_70006E
.rdata:0068477C                 dd offset unk_740075
.rdata:00684780                 dd offset loc_5F0031
.rdata:00684784 a3_dll:
.rdata:00684784                 unicode 0, <3.dll>,0
.rdata:00684790 qword_684790    dq 40EFFFE000000000h    ; DATA XREF: sub_43E540+185r
.rdata:00684790                                         ; sub_503360+42Br ...
.rdata:00684798 qword_684798    dq 3F00000000000000h    ; DATA XREF: sub_43E540+110r
.rdata:00684798                                         ; sub_5A38F0:loc_5A3A6Er ...
.rdata:006847A0 qword_6847A0    dq 3F70000000000000h    ; DATA XREF: sub_43E540+D0r
.rdata:006847A0                                         ; sub_49BF60+92r ...
.rdata:006847A8 aXbox360Contr_5 db 'Xbox 360 Controller (XInput DRUM KIT)',0
.rdata:006847A8                                         ; DATA XREF: sub_43E710+D7o
.rdata:006847CE                 align 10h
.rdata:006847D0 aXbox360Contr_4 db 'Xbox 360 Controller (XInput GUITAR)',0
.rdata:006847D0                                         ; DATA XREF: sub_43E710+C0o
.rdata:006847F4 aXbox360Contr_3 db 'Xbox 360 Controller (XInput DANCE PAD)',0
.rdata:006847F4                                         ; DATA XREF: sub_43E710+AAo
.rdata:0068481B                 align 4
.rdata:0068481C aXbox360Contr_2 db 'Xbox 360 Controller (XInput FLIGHT STICK)',0
.rdata:0068481C                                         ; DATA XREF: sub_43E710+93o
.rdata:00684846                 align 4
.rdata:00684848 aXbox360Contr_1 db 'Xbox 360 Controller (XInput ARCADE STICK)',0
.rdata:00684848                                         ; DATA XREF: sub_43E710+7Co
.rdata:00684872                 align 4
.rdata:00684874 aXbox360Contr_0 db 'Xbox 360 Controller (XInput WHEEL)',0
.rdata:00684874                                         ; DATA XREF: sub_43E710+66o
.rdata:00684897                 align 4
.rdata:00684898 aXbox360Control db 'Xbox 360 Controller (XInput STANDARD GAMEPAD)',0
.rdata:00684898                                         ; DATA XREF: sub_43E710+4Fo
.rdata:006848C6                 align 4
.rdata:006848C8 aXinput1_3_dllN db '!!!!! xinput1_3.dll not found. Gamepad support disabled',0
.rdata:006848C8                                         ; DATA XREF: sub_43E970+1Co
```

File: ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp (hunk 3 above)
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp]: This is taken care of.

File: ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSjoystick.cpp (hunk 1 above)
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSjoystick.cpp]: This is taken care of.

File: ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSjoystick.cpp (hunk 2 above)
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSjoystick.cpp]: Do you want that inlined in the enigma_user namespace or the enigma namespace?

**Comment by JoshDreamland** [comment]
I seriously doubt those are the only Strings GM supports identifying a controller with. They may be custom strings for controllers they had on hand and could prepare a customized string for, but I was able to identify my Logitech controller with these joystick APIs just fine, so more are available.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp [hunk 4]
```diff
@@ -24,231 +23,230 @@
...
 
-    // Vibrate the controller
-    XInputSetState(device, &vibration);
+  XInputSetState(device, &vibration);
 }
-	
+
 int gamepad_axis_count(int device) {
-	return 2;
+  return 4;
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp]: This isn't the right answer in any universe; query the API for the axis count.

Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp]: When it comes to XInput Josh, 4 controllers is the limit.
http://gaming.stackexchange.com/a/196367

For DirectInput you can have as many as you want.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp [hunk 5]
```diff
@@ -24,231 +23,230 @@
...
+    } else {
+      return ret / 32768;
+    }
+  } else {
+    return 0;
+  }
+}
+
+WORD digitalButtons[20] = {
+    XINPUT_GAMEPAD_A, XINPUT_GAMEPAD_X, XINPUT_GAMEPAD_B, XINPUT_GAMEPAD_Y,
```
Comment (SuperRiderTH) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp]: XINPUT_GAMEPAD_X and XINPUT_GAMEPAD_B need to be switched. gp_face2 is B, and gp_face3 is X.

Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp]: Thank you for pointing it out, I will check it out before merging this.

**Comment by JoshDreamland** [comment]
The diff on this PR is seriously FUBAR. I imagine you did a lot of work on this branch. Do a pull and see if that makes the diff make sense. There's only one file with conflicts (`roomsystem.cpp`); I suggest reverting it if it looks hairy.

**Comment by RobertBColton** [comment]
Uhm I fixed XInput in this one and also cleaned whitespace/indentation in joysticks and the Windows exe.

**Comment by JoshDreamland** [comment]
What I'm saying is, looking at the diff, I can't imagine pulling this. I'm hoping the diff is just outdated, because it's showing massive changes that look like reverts and regressions. Try merging in master and fixing the merge conflicts with Meld.

**Comment by RobertBColton** [comment]
I was really just going to redo them in new pull requests Josh. I just don't want to lose this ticket.

**Comment by RobertBColton** [comment]
Closing this for a number of reasons, and because it's crusty as fuck.

1) The Windows joystick formatting issues are covered by #1223
2) Installed Java can just go to hell.
3) This has too many conflicts.
4) I can just redo the gamepad functions anyway, I know what needs to be done, and I'm very familiar with XInput.

We don't need this one anymore.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp (hunk 5 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp]: I took care of it in my updated XInput pull request that I am creating: https://github.com/enigma-dev/enigma-dev/commit/4d4f11f616ac682679f2213affb6453c3da3c02c

You were correct, thanks for the help.

**Pull request closed without merging**